### DISCUSSION TITLE
How to get the result of an approval?

### DISCUSSION THREAD

**Discussion by wsw70** [windmill-labs/windmill#2763] [Questions]
I managed to write my first approval in Python, and then accept or reject it 🎉

I now would like to react differently depending on the result, but the exact flow is not clear to me.

As an example, I have a single approver and the approval step is:

```python
import wmill

def main(x: str):
    urls = wmill.get_resume_urls()
    print(urls)
    return x
```

This allows me to grab the JSON structure with the approve and reject URLs. When I call these URLs, the flow completes successfully (approved), or with an error (rejected).

What I would like to understand is

- where exactly the script is suspended
- and how to retrieve the result of the approval

**where the script is suspended**

My approval will be rather typical: at the place where I `print(urls)` above, I would actually send an email with some content. My understanding is that the script will go past `return x`, and stop right before handing over the flow (and awaiting for the web call). Is that correct?

**how to retrieve the result of the approval**

If the above is correct, how can I get the results of the approval? Is this based on a success/failure of the step? In which case a success would continue the script - but how can I catch the failure?

I am hoping for two branches - one for the approved step, and the other for the failed one:

![image](https://github.com/windmill-labs/windmill/assets/1345886/75354503-ff47-457a-8b4f-170868dd06e7)

The "resume" URL goes as expected. The "cancel" URL ends the whole flow with

```
Flow job canceledUnexpected error during job execution:
Object {
    "message": String("Job canceled: approval request disapproved by "),
    "name": String("Canceled"),
    "reason": String("approval request disapproved"),
    "canceler": String(""),
}
```

I would like to be able to redirect the flow when the above happens

**Comment by rubenfiszel** [comment]
You can get the result of the approval at the next step (using  `resumes = list of all payloads` or `resume = first payload`. The approval step form is configurable in Advance -> Suspend -> Form

We recommend adding a selectbox or similar and treat it in a branchone to have the desired effect, cancel will always stop a flow at that step. We may make removing the cancel button a configuration later.

This documentation is pretty complete at https://www.windmill.dev/docs/flows/flow_approval

**Comment by wsw70** [comment]
Aaahhhh, now I get it - the "actual approval" is not driven by the `cancel` and `resume` URLs, but by the form. Thank you for pointing that out, I of course read the documentation you mention but could not make it out that the form is central to the approval.

Can that form be edited? I could not find a link to the UI editor.

If it cannot, I will go for an email that only has the "resume" URL (as "approved") and handle the non-approval manually. I saw that the "resume" URL points to a blank page, I guess that this cannot be edited either (even with a simple message "approved" or something like that)?

**Reply by rubenfiszel** [reply]
The form can be edited:
Advanced -> Suspend -> Form

you can there add any properties of any type. We recommend using the approval page rather than the resume urls since most email providers will pre-GET the links in your email thus triggering the approval of the flow

**Comment by wsw70** [comment]
> The form can be edited: Advanced -> Suspend -> Form

Sorry, I was not clear. I meant "can it be edited *completely*". That is create my own page with form elements (an approval switch for instance, and some text, and an extra comment text field, ...), the same way as auto-generated forms can be edited.

> most email providers will pre-GET the links

I am aware of that, thank you for reminding that point. I would have control over the MTA (especially if we go for the Enterprise license)

**Reply by rubenfiszel** [reply]
Absolutely, the form is actually only a UI config and has nothing to do with what the API actually support:

Anything that is passed as body of this POST: https://app.windmill.dev/openapi.html#/operations/resumeSuspendedJobPost (which is the URL you get in resumeUrl) will be passed as the resume payload. 

We merely provide reasonable default implementation for it with the approval page but it's fully customizable.

**Reply by rubenfiszel** [reply]
to note, we also support arbitrary encoding with GET method but the payload need to be encoded in the query with base64: https://app.windmill.dev/openapi.html#/operations/resumeSuspendedJobGet
